File: go-frontend/internal/semantic/source_map.go
package semantic

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"

	"freehold-go-frontend/internal/ast"
	"freehold-go-frontend/internal/token"
)

const SourceMapSchemaV0 = "fh-source-map-v0"

type sourceMapDocument struct {
	Schema          string            `json:"schema"`
	LanguageVersion string            `json:"language_version"`
	Purpose         string            `json:"purpose"`
	EntryModule     string            `json:"entry_module"`
	ModuleOrder     []string          `json:"module_order"`
	Modules         []sourceMapModule `json:"modules"`
}

type sourceMapModule struct {
	Name         string        `json:"name"`
	SourceFile   string        `json:"source_file"`
	SourceSHA256 string        `json:"source_sha256"`
	LineCount    int           `json:"line_count"`
	SourceLines  []string      `json:"source_lines"`
	AST          sourceMapNode `json:"ast"`
}

type sourceMapNode struct {
	ID       string          `json:"id"`
	ParentID *string         `json:"parent_id"`
	Role     string          `json:"role"`
	Index    int             `json:"index"`
	Kind     string          `json:"kind"`
	Span     sourceMapSpan   `json:"span"`
	Extra    map[string]any  `json:"-"`
	Children []sourceMapNode `json:"children"`
}

type sourceMapSpan struct {
	Line     *int   `json:"line"`
	Column   *int   `json:"column"`
	LineText string `json:"line_text"`
}

type sourceChild struct {
	role  string
	index int
	node  any
}

func (n sourceMapNode) MarshalJSON() ([]byte, error) {
	item := map[string]any{
		"id":        n.ID,
		"parent_id": n.ParentID,
		"role":      n.Role,
		"index":     n.Index,
		"kind":      n.Kind,
		"span":      n.Span,
		"children":  n.Children,
	}
	for key, value := range n.Extra {
		item[key] = value
	}
	return json.Marshal(item)
}

func ExportSourceMapJSON(entryFile string) (string, error) {
	project, err := LoadProject(entryFile)
	if err != nil {
		return "", err
	}
	if project == nil || project.Entry == nil {
		return "", fmt.Errorf("failed to load project")
	}

	moduleNames := project.ModuleNames()
	modules := make([]sourceMapModule, 0, len(moduleNames))
	for _, moduleName := range moduleNames {
		module, ok := project.Modules[moduleName]
		if !ok || module == nil {
			continue
		}
		path := project.Files[moduleName]
		exported, err := exportSourceMapModule(moduleName, path, module)
		if err != nil {
			return "", err
		}
		modules = append(modules, exported)
	}

	doc := sourceMapDocument{
		Schema:          SourceMapSchemaV0,
		LanguageVersion: FreeholdLanguageVersion,
		Purpose:         "source reconstruction sidecar for FH-IR",
		EntryModule:     project.Entry.Name,
		ModuleOrder:     moduleNames,
		Modules:         modules,
	}

	var buffer bytes.Buffer
	encoder := json.NewEncoder(&buffer)
	encoder.SetEscapeHTML(false)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(doc); err != nil {
		return "", err
	}
	return buffer.String(), nil
}

func exportSourceMapModule(moduleName string, path string, module *ast.Module) (sourceMapModule, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return sourceMapModule{}, err
	}
	source := normalizeSourceText(string(content))
	lines := splitSourceLines(source)
	digest := sha256.Sum256([]byte(source))
	return sourceMapModule{
		Name:         moduleName,
		SourceFile:   path,
		SourceSHA256: hex.EncodeToString(digest[:]),
		LineCount:    len(lines),
		SourceLines:  lines,
		AST:          exportSourceNode(module, moduleName, "module", nil, "module", 0, lines),
	}, nil
}

func splitSourceLines(source string) []string {
	source = normalizeSourceText(source)
	if source == "" {
		return []string{}
	}
	lines := strings.Split(source, "\n")
	if len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}
	return lines
}

func normalizeSourceText(source string) string {
	source = strings.ReplaceAll(source, "\r\n", "\n")
	return strings.ReplaceAll(source, "\r", "\n")
}

func exportSourceNode(node any, moduleName string, path string, parentID *string, role string, index int, lines []string) sourceMapNode {
	id := moduleName + ":" + path
	children := sourceChildren(node)
	exportedChildren := make([]sourceMapNode, 0, len(children))
	for _, child := range children {
		childPath := fmt.Sprintf("%s/%s[%d]", path, child.role, child.index)
		exportedChildren = append(exportedChildren, exportSourceNode(child.node, moduleName, childPath, &id, child.role, child.index, lines))
	}
	return sourceMapNode{
		ID:       id,
		ParentID: parentID,
		Role:     role,
		Index:    index,
		Kind:     sourceKind(node),
		Span:     sourceSpan(sourcePos(node), lines),
		Extra:    sourceSummary(node),
		Children: exportedChildren,
	}
}

func sourceSpan(pos token.Position, lines []string) sourceMapSpan {
	if pos.Line <= 0 {
		return sourceMapSpan{Line: nil, Column: nil, LineText: ""}
	}
	line := pos.Line
	column := pos.Column
	lineText := ""
	if line > 0 && line <= len(lines) {
		lineText = lines[line-1]
	}
	return sourceMapSpan{Line: &line, Column: &column, LineText: lineText}
}

func sourceKind(node any) string {
	switch node.(type) {
	case *ast.Module, ast.Module:
		return "Program"
	case ast.TypeDecl:
		value := node.(ast.TypeDecl)
		if value.Base == "record" {
			return "RecordTypeDecl"
		}
		return "TypeDecl"
	case ast.FunctionDecl, ast.ProcedureDecl:
		return "RoutineDecl"
	case ast.Param:
		return "Param"
	case ast.AssignmentStmt:
		value := node.(ast.AssignmentStmt)
		if fieldPathFromExpr(value.Target) != nil {
			return "FieldAssignStmt"
		}
		return "AssignStmt"
	case ast.IdentifierExpr:
		return "VarExpr"
	case ast.NamedArgumentExpr, ast.RecordField:
		return "NamedArg"
	case ast.OkExpr:
		return "ReturnOk"
	case ast.ErrorExpr:
		return "ReturnError"
	default:
		name := fmt.Sprintf("%T", node)
		if strings.HasPrefix(name, "ast.") {
			return strings.TrimPrefix(name, "ast.")
		}
		return name
	}
}

func sourcePos(node any) token.Position {
	switch value := node.(type) {
	case *ast.Module:
		return value.Pos
	case ast.Module:
		return value.Pos
	case ast.ImportDecl:
		return value.Pos
	case ast.TypeDecl:
		return value.Pos
	case ast.ErrorDecl:
		return value.Pos
	case ast.ServiceDecl:
		return value.Pos
	case ast.RpcDecl:
		return value.Pos
	case ast.FunctionDecl:
		return value.Pos
	case ast.ProcedureDecl:
		return value.Pos
	case ast.AbortClause:
		return value.Pos
	case ast.Param:
		return value.Pos
	case ast.GlobalSpec:
		return value.Pos
	case ast.DependsSpec:
		return value.Pos
	case ast.ReturnStmt:
		return value.Pos
	case ast.AbortStmt:
		return value.Pos
	case ast.CheckStmt:
		return value.Pos
	case ast.LetStmt:
		return value.Pos
	case ast.AssignmentStmt:
		return value.Pos
	case ast.CallStmt:
		return value.Pos
	case ast.IfStmt:
		return value.Pos
	case ast.WhileStmt:
		return value.Pos
	case ast.CaseStmt:
		return value.Pos
	case ast.ScopeStmt:
		return value.Pos
	case ast.CaseBranch:
		return value.Pos
	case ast.IdentifierExpr:
		return value.Pos
	case ast.FieldAccessExpr:
		return value.Pos
	case ast.IndexExpr:
		return value.Pos
	case ast.ArrayLiteralExpr:
		return value.Pos
	case ast.CallExpr:
		return value.Pos
	case ast.AwaitExpr:
		return value.Pos
	case ast.NamedArgumentExpr:
		return value.Pos
	case ast.RecordLiteralExpr:
		return value.Pos
	case ast.RecordField:
		return value.Pos
	case ast.NumberExpr:
		return value.Pos
	case ast.StringExpr:
		return value.Pos
	case ast.OkExpr:
		return value.Pos
	case ast.ErrorExpr:
		return value.Pos
	case ast.UnaryExpr:
		return value.Pos
	case ast.BinaryExpr:
		return value.Pos
	case ast.ForAllExpr:
		return value.Pos
	case ast.ExistsExpr:
		return value.Pos
	default:
		return token.Position{}
	}
}

func sourceSummary(node any) map[string]any {
	item := map[string]any{}
	switch value := node.(type) {
	case *ast.Module:
		item["name"] = value.Name
	case ast.Module:
		item["name"] = value.Name
	case ast.ImportDecl:
		item["module"] = value.Module
		item["exposing"] = value.Exposing
	case ast.TypeDecl:
		item["name"] = value.Name
		if value.Base == "record" {
			item["type_params"] = nonNilStrings(value.TypeParams)
		} else {
			item["base"] = value.Base
			if value.Range != nil {
				item["min_value"] = numericOrString(value.Range.Min)
				item["max_value"] = numericOrString(value.Range.Max)
			}
		}
	case ast.Param:
		item["name"] = value.Name
		item["type"] = value.Type
		if value.ProtoID != nil {
			item["proto_id"] = *value.ProtoID
		}
	case ast.ErrorDecl:
		item["name"] = value.Name
	case ast.ServiceDecl:
		item["name"] = value.Name
	case ast.RpcDecl:
		item["name"] = value.Name
		item["request_name"] = value.RequestName
		item["request_type"] = value.RequestType
		item["response_type"] = value.ResponseType
		item["request_stream"] = value.RequestStream
		item["response_stream"] = value.ResponseStream
	case ast.FunctionDecl:
		item["name"] = value.Name
		item["routine_kind"] = "function"
		item["is_async"] = value.IsAsync
		item["type_params"] = nonNilStrings(value.TypeParams)
		item["return_type"] = value.ReturnType
	case ast.ProcedureDecl:
		item["name"] = value.Name
		item["routine_kind"] = "procedure"
		item["is_async"] = false
		item["type_params"] = []string{}
		item["return_type"] = "Void"
	case ast.GlobalSpec:
		item["name"] = value.Name
		if value.Mode != nil {
			item["mode"] = *value.Mode
		} else {
			item["mode"] = nil
		}
	case ast.DependsSpec:
		item["target"] = value.Target
		item["sources"] = value.Sources
	case ast.AbortClause:
		item["error"] = value.Error
	case ast.LetStmt:
		item["name"] = value.Name
		item["type"] = value.Type
	case ast.AssignmentStmt:
		if path := fieldPathFromExpr(value.Target); path != nil {
			item["path"] = path
		} else if name := identifierName(value.Target); name != "" {
			item["name"] = name
		}
	case ast.AbortStmt:
		item["error"] = value.Error
	case ast.CallStmt:
		if call, ok := value.Call.(ast.CallExpr); ok {
			if name := callNameForSource(call.Callee); name != "" {
				item["name"] = name
			}
			item["type_args"] = nonNilStrings(call.TypeArgs)
		}
	case ast.ScopeStmt:
		item["name"] = value.Name
	case ast.IdentifierExpr:
		item["name"] = value.Name
	case ast.FieldAccessExpr:
		item["path"] = fieldPathFromExpr(value)
	case ast.CallExpr:
		item["name"] = callNameForSource(value.Callee)
		item["type_args"] = nonNilStrings(value.TypeArgs)
	case ast.NamedArgumentExpr:
		item["name"] = value.Name
	case ast.RecordLiteralExpr:
		item["type"] = value.Type
	case ast.RecordField:
		item["name"] = value.Name
	case ast.IndexExpr:
		item["name"] = callNameForSource(value.Array)
	case ast.NumberExpr:
		item["value"] = numericOrString(value.Value)
	case ast.StringExpr:
		item["value"] = value.Value
	case ast.OkExpr:
	case ast.ErrorExpr:
		item["error"] = value.Name
	case ast.UnaryExpr:
		item["op"] = value.Op
	case ast.BinaryExpr:
		item["op"] = value.Op
	case ast.ForAllExpr:
		item["var_name"] = value.VarName
	case ast.ExistsExpr:
		item["var_name"] = value.VarName
	}
	return item
}

func sourceChildren(node any) []sourceChild {
	var children []sourceChild
	add := func(role string, nodes []any) {
		for index, child := range nodes {
			if child != nil {
				children = append(children, sourceChild{role: role, index: index, node: child})
			}
		}
	}
	one := func(role string, node any) {
		if node != nil {
			children = append(children, sourceChild{role: role, index: 0, node: node})
		}
	}

	switch value := node.(type) {
	case *ast.Module:
		add("declarations", declsToAny(value.Declarations))
	case ast.Module:
		add("declarations", declsToAny(value.Declarations))
	case ast.TypeDecl:
		if value.Base == "record" {
			add("fields", paramsToAny(value.Fields))
		}
	case ast.ServiceDecl:
		add("rpcs", rpcsToAny(value.Rpcs))
	case ast.FunctionDecl:
		add("params", paramsToAny(value.Params))
		add("globals", globalsToAny(value.GlobalSpecs))
		add("depends", dependsToAny(value.DependsSpecs))
		add("requires", exprsToAny(value.Requires))
		add("aborts", abortsToAny(value.Aborts))
		add("ensures", exprsToAny(value.Ensures))
		add("body", stmtsToAny(value.Body))
	case ast.ProcedureDecl:
		add("params", paramsToAny(value.Params))
		add("globals", globalsToAny(value.GlobalSpecs))
		add("depends", dependsToAny(value.DependsSpecs))
		add("requires", exprsToAny(value.Requires))
		add("aborts", abortsToAny(value.Aborts))
		add("ensures", exprsToAny(value.Ensures))
		add("body", stmtsToAny(value.Body))
	case ast.AbortClause:
		one("condition", value.Condition)
	case ast.ReturnStmt:
		one("value", value.Value)
	case ast.CheckStmt:
		one("condition", value.Condition)
	case ast.LetStmt:
		one("value", value.Value)
	case ast.AssignmentStmt:
		one("value", value.Value)
	case ast.CallStmt:
		if call, ok := value.Call.(ast.CallExpr); ok {
			add("args", exprsToAny(call.Arguments))
		} else {
			one("call", value.Call)
		}
	case ast.IfStmt:
		one("condition", value.Condition)
		add("then_body", stmtsToAny(value.ThenBody))
		add("else_body", stmtsToAny(value.ElseBody))
	case ast.WhileStmt:
		one("condition", value.Condition)
		add("invariants", exprsToAny(value.Invariants))
		one("variant", value.Variant)
		add("body", stmtsToAny(value.Body))
	case ast.CaseStmt:
		one("value", value.Value)
		add("branches", caseBranchesToAny(value.When))
		add("default_body", stmtsToAny(value.Default))
	case ast.CaseBranch:
		one("value", value.Value)
		add("body", stmtsToAny(value.Body))
	case ast.ScopeStmt:
		add("spawn_body", stmtsToAny(value.SpawnBody))
		add("join_body", stmtsToAny(value.JoinBody))
		add("result_body", stmtsToAny(value.ResultBody))
	case ast.FieldAccessExpr:
		one("object", value.Object)
	case ast.IndexExpr:
		one("array", value.Array)
		one("index", value.Index)
	case ast.ArrayLiteralExpr:
		add("items", exprsToAny(value.Elements))
	case ast.CallExpr:
		add("args", exprsToAny(value.Arguments))
	case ast.AwaitExpr:
		one("value", value.Value)
	case ast.NamedArgumentExpr:
		one("value", value.Value)
	case ast.RecordLiteralExpr:
		add("args", recordFieldsToAny(value.Fields))
	case ast.RecordField:
		one("value", value.Value)
	case ast.OkExpr:
		one("value", value.Value)
	case ast.UnaryExpr:
		one("value", value.Value)
	case ast.BinaryExpr:
		one("left", value.Left)
		one("right", value.Right)
	case ast.ForAllExpr:
		one("lower", value.Lower)
		one("upper", value.Upper)
		one("expr", value.Expr)
	case ast.ExistsExpr:
		one("lower", value.Lower)
		one("upper", value.Upper)
		one("expr", value.Expr)
	}
	return children
}

func declsToAny(values []ast.Decl) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func paramsToAny(values []ast.Param) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func rpcsToAny(values []ast.RpcDecl) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func globalsToAny(values []ast.GlobalSpec) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func dependsToAny(values []ast.DependsSpec) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func abortsToAny(values []ast.AbortClause) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func stmtsToAny(values []ast.Stmt) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func exprsToAny(values []ast.Expr) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func caseBranchesToAny(values []ast.CaseBranch) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func recordFieldsToAny(values []ast.RecordField) []any {
	items := make([]any, len(values))
	for i, value := range values {
		items[i] = value
	}
	return items
}

func nonNilStrings(values []string) []string {
	if values == nil {
		return []string{}
	}
	return values
}

func numericOrString(value string) any {
	if strings.Contains(value, ".") {
		return value
	}
	var parsed int
	if _, err := fmt.Sscanf(value, "%d", &parsed); err == nil {
		return parsed
	}
	return value
}

func identifierName(expr ast.Expr) string {
	if value, ok := expr.(ast.IdentifierExpr); ok {
		return value.Name
	}
	return ""
}

func callNameForSource(expr ast.Expr) string {
	switch value := expr.(type) {
	case ast.IdentifierExpr:
		return value.Name
	case ast.FieldAccessExpr:
		path := fieldPathFromExpr(value)
		if len(path) > 0 {
			return strings.Join(path, ".")
		}
	}
	return ""
}

func fieldPathFromExpr(expr ast.Expr) []string {
	switch value := expr.(type) {
	case ast.IdentifierExpr:
		return []string{value.Name}
	case ast.FieldAccessExpr:
		prefix := fieldPathFromExpr(value.Object)
		if prefix == nil {
			return nil
		}
		return append(prefix, value.Field)
	}
	return nil
}

func sortedSourceMapKeys(item map[string]any) []string {
	keys := make([]string, 0, len(item))
	for key := range item {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

File: freehold/core/source_map.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any

from freehold.core.ast import (
    AbortClause,
    AbortStmt,
    ArrayLiteralExpr,
    AssignStmt,
    AwaitExpr,
    BinaryExpr,
    BoolExpr,
    CallExpr,
    CallStmt,
    CaseBranch,
    CaseStmt,
    CheckStmt,
    DoubleExpr,
    ErrorDecl,
    ExistsExpr,
    FieldAccessExpr,
    FieldAssignStmt,
    ForAllExpr,
    GlobalSpec,
    IfStmt,
    ImportDecl,
    IndexExpr,
    IndexedFieldAccessExpr,
    LetStmt,
    MapEntry,
    MapLiteralExpr,
    NamedArg,
    NumberExpr,
    Param,
    PatternBranch,
    PatternExpr,
    Program,
    RecordField,
    RecordLiteralExpr,
    RecordTypeDecl,
    ReturnError,
    ReturnOk,
    ReturnPlain,
    ReturnStmt,
    RoutineDecl,
    ScopeStmt,
    ServiceDecl,
    SourcePos,
    StringExpr,
    TypeDecl,
    UnaryExpr,
    VarExpr,
    WhileStmt,
    type_to_string,
)
from freehold.core.go_codegen import GO_RUNTIME_MODULE_EXPORTS
from freehold.core.module_resolver import ModuleResolver, ResolvedModule

SOURCE_MAP_SCHEMA = "fh-source-map-v0"
FREEHOLD_LANGUAGE_VERSION = "freehold-v1"


def export_source_map_json(entry_file: str | Path) -> str:
    resolver = ModuleResolver(runtime_modules=GO_RUNTIME_MODULE_EXPORTS)
    resolver.resolve_entry(entry_file)
    if resolver.entry is None:
        raise RuntimeError("module resolver did not produce an entry module")
    return json.dumps(export_resolved_source_map(resolver.entry.name, resolver.resolved), indent=2, ensure_ascii=False) + "\n"


def export_resolved_source_map(entry_module_name: str, resolved_modules: dict[str, ResolvedModule]) -> dict[str, Any]:
    modules = [export_module_source_map(resolved_modules[name]) for name in sorted(resolved_modules)]
    return {
        "schema": SOURCE_MAP_SCHEMA,
        "language_version": FREEHOLD_LANGUAGE_VERSION,
        "purpose": "source reconstruction sidecar for FH-IR",
        "entry_module": entry_module_name,
        "module_order": [module["name"] for module in modules],
        "modules": modules,
    }


def export_module_source_map(resolved: ResolvedModule) -> dict[str, Any]:
    source = resolved.path.read_text(encoding="utf-8")
    lines = source.splitlines()
    return {
        "name": resolved.name,
        "source_file": str(resolved.path),
        "source_sha256": hashlib.sha256(source.encode("utf-8")).hexdigest(),
        "line_count": len(lines),
        "source_lines": lines,
        "ast": export_node(resolved.ast, resolved.name, "module", None, "module", 0, lines),
    }


def export_node(
    node: Any,
    module_name: str,
    path: str,
    parent_id: str | None,
    role: str,
    index: int,
    lines: list[str],
) -> dict[str, Any]:
    node_id = f"{module_name}:{path}"
    item: dict[str, Any] = {
        "id": node_id,
        "parent_id": parent_id,
        "role": role,
        "index": index,
        "kind": type(node).__name__,
        "span": export_span(getattr(node, "pos", None), lines),
    }
    item.update(node_summary(node))
    children = list(iter_children(node))
    if children:
        item["children"] = [
            export_node(child, module_name, f"{path}/{child_role}[{child_index}]", node_id, child_role, child_index, lines)
            for child_role, child_index, child in children
        ]
    else:
        item["children"] = []
    return item


def export_span(pos: SourcePos | None, lines: list[str]) -> dict[str, Any]:
    if pos is None or pos.line is None:
        return {"line": None, "column": None, "line_text": ""}
    line_text = lines[pos.line - 1] if 0 < pos.line <= len(lines) else ""
    return {"line": pos.line, "column": pos.column, "line_text": line_text}


def node_summary(node: Any) -> dict[str, Any]:
    if isinstance(node, Program):
        return {"name": node.module_name}
    if isinstance(node, ImportDecl):
        return {"module": node.module_name, "exposing": list(node.exposing)}
    if isinstance(node, TypeDecl):
        return {"name": node.name, "base": node.base, "min_value": node.min_value, "max_value": node.max_value}
    if isinstance(node, RecordTypeDecl):
        return {"name": node.name, "type_params": list(node.type_params or [])}
    if isinstance(node, RecordField):
        item = {"name": node.name, "type": node.type_name}
        if node.proto_id is not None:
            item["proto_id"] = node.proto_id
        if node.json_name is not None:
            item["json_name"] = node.json_name
        return item
    if isinstance(node, ErrorDecl):
        return {"name": node.name}
    if isinstance(node, ServiceDecl):
        return {"name": node.name}
    if isinstance(node, RoutineDecl):
        return {
            "name": node.name,
            "routine_kind": node.kind,
            "is_async": node.is_async,
            "type_params": list(node.type_params or []),
            "return_type": type_to_string(node.return_type),
        }
    if isinstance(node, Param):
        return {"name": node.name, "type": node.type_name}
    if isinstance(node, GlobalSpec):
        return {"name": node.name, "mode": node.mode}
    if isinstance(node, AbortClause):
        return {"error": node.error_name}
    if isinstance(node, LetStmt):
        return {"name": node.name, "type": type_to_string(node.type_ref)}
    if isinstance(node, AssignStmt):
        return {"name": node.name}
    if isinstance(node, FieldAssignStmt):
        return {"path": list(node.path)}
    if isinstance(node, AbortStmt):
        return {"error": node.error_name}
    if isinstance(node, CallStmt):
        return {"name": node.name, "type_args": list(node.type_args or [])}
    if isinstance(node, ScopeStmt):
        return {"name": node.name}
    if isinstance(node, ReturnError):
        return {"error": node.error_name}
    if isinstance(node, StringExpr):
        return {"value": node.value}
    if isinstance(node, NumberExpr):
        return {"value": node.value}
    if isinstance(node, DoubleExpr):
        return {"value": node.value}
    if isinstance(node, BoolExpr):
        return {"value": node.value}
    if isinstance(node, VarExpr):
        return {"name": node.name}
    if isinstance(node, FieldAccessExpr):
        return {"path": list(node.path)}
    if isinstance(node, CallExpr):
        return {"name": node.name, "type_args": list(node.type_args or [])}
    if isinstance(node, NamedArg):
        return {"name": node.name}
    if isinstance(node, RecordLiteralExpr):
        return {"type": node.type_name}
    if isinstance(node, MapLiteralExpr):
        return {"type": node.type_name}
    if isinstance(node, MapEntry):
        return {"key": node.key}
    if isinstance(node, IndexExpr):
        return {"name": node.name}
    if isinstance(node, IndexedFieldAccessExpr):
        return {"name": node.name, "fields": list(node.fields)}
    if isinstance(node, UnaryExpr):
        return {"op": node.op}
    if isinstance(node, BinaryExpr):
        return {"op": node.op}
    if isinstance(node, ForAllExpr):
        return {"var_name": node.var_name}
    if isinstance(node, ExistsExpr):
        return {"var_name": node.var_name}
    return {}


def iter_children(node: Any) -> list[tuple[str, int, Any]]:
    children: list[tuple[str, int, Any]] = []

    def add(role: str, values: list[Any] | tuple[Any, ...] | None) -> None:
        if not values:
            return
        for child_index, child in enumerate(values):
            if child is not None:
                children.append((role, child_index, child))

    def one(role: str, value: Any | None) -> None:
        if value is not None:
            children.append((role, 0, value))

    if isinstance(node, Program):
        add("imports", node.imports or [])
        add("declarations", node.declarations)
    elif isinstance(node, RecordTypeDecl):
        add("fields", node.fields)
    elif isinstance(node, ServiceDecl):
        add("rpcs", node.rpcs)
    elif isinstance(node, RoutineDecl):
        add("params", node.params)
        add("globals", node.global_specs or [])
        add("requires", node.requires)
        add("aborts", node.aborts)
        add("ensures", node.ensures)
        add("body", node.body)
    elif isinstance(node, AbortClause):
        one("condition", node.condition)
    elif isinstance(node, LetStmt):
        one("value", node.expr)
    elif isinstance(node, AssignStmt):
        one("value", node.expr)
    elif isinstance(node, FieldAssignStmt):
        one("value", node.expr)
    elif isinstance(node, ReturnStmt):
        one("value", node.value)
    elif isinstance(node, ReturnPlain):
        one("value", node.expr)
    elif isinstance(node, ReturnOk):
        one("value", node.expr)
    elif isinstance(node, CheckStmt):
        one("condition", node.expr)
    elif isinstance(node, CallStmt):
        add("args", node.args)
    elif isinstance(node, IfStmt):
        one("condition", node.condition)
        add("then_body", node.then_body)
        add("else_body", node.else_body)
    elif isinstance(node, WhileStmt):
        one("condition", node.condition)
        add("invariants", node.invariants)
        one("variant", node.variant)
        add("body", node.body)
    elif isinstance(node, CaseStmt):
        one("value", node.expr)
        add("branches", node.branches)
        add("default_body", node.default_body)
    elif isinstance(node, CaseBranch):
        one("value", node.value)
        add("body", node.body)
    elif isinstance(node, PatternBranch):
        one("guard", node.guard)
        add("body", node.body)
    elif isinstance(node, ScopeStmt):
        add("spawn_body", node.spawn_body)
        add("join_body", node.join_body)
        add("result_body", node.result_body)
    elif isinstance(node, CallExpr):
        add("args", node.args)
        one("invariant", node.invariant)
    elif isinstance(node, AwaitExpr):
        one("value", node.expr)
    elif isinstance(node, NamedArg):
        one("value", node.expr)
    elif isinstance(node, RecordLiteralExpr):
        add("args", node.args)
    elif isinstance(node, MapLiteralExpr):
        add("entries", node.entries)
    elif isinstance(node, MapEntry):
        one("expr", node.expr)
    elif isinstance(node, ArrayLiteralExpr):
        add("items", node.items)
    elif isinstance(node, IndexExpr):
        one("index", node.index)
    elif isinstance(node, IndexedFieldAccessExpr):
        one("index", node.index)
    elif isinstance(node, UnaryExpr):
        one("value", node.expr)
    elif isinstance(node, BinaryExpr):
        one("left", node.left)
        one("right", node.right)
    elif isinstance(node, ForAllExpr):
        one("lower", node.lower)
        one("upper", node.upper)
        one("expr", node.expr)
    elif isinstance(node, ExistsExpr):
        one("lower", node.lower)
        one("upper", node.upper)
        one("expr", node.expr)
    return children
